### DISCUSSION TITLE
Consider targeting wgpu as lowest common denominator for graphics programming

### DISCUSSION THREAD

**Discussion by erlend-sh** [vlang/v#11670] [Ideas]
A great argument is made in favor of wgpu by @kvark here:

http://kvark.github.io/web/gpu/native/2020/05/03/point-of-webgpu-native.html

> If you are developing an application today, the choice of the tech stack looks something like this:
> 
> - target OpenGL only, accept the performance hits and bugs, potentially route via Angle.
> - target a graphics library like bgfx or sokol-gfx, accept working without a spec, also be ready to fix bugs and submit patches.
> - target Vulkan Portability, accept the complexity.
> - develop your own graphics backends on selected platforms, accept the development costs.
> - pick an engine like Unity/Unreal/Lumberyard that gives you portable graphics, accept all the features that come with it.
> 
> Today’s engines are pretty good! They are powerful and cheap to use. But I think, a lot of appeal to them was driven by the lack of proper solutions in the space below. Dealing with different graphics APIs on platforms is hard. Leaving it to a third-party library means putting a lot of trust in it.
> 
> If WebGPU on native becomes practical, it would be strictly superior to (1) and (2). It will be supported by an actual specification, a rich conformance test suite, big corporations, and tested to death by the use in browsers. The trade-offs it provides would make it a solid choice even for developers who’d otherwise go for (3), (4), and (5), but not all of them of course.
> 
> I can see WebGPU on native being a go-to choice for amateur developers, students, indie professionals, mobile game studios, and many other groups. It could be the default GPU API, if it can deliver on its promises of safety, performance, and portability. We have a lot of interest and early adopters, as well as big forces in motion to make this real.

**Comment by JalonSolov** [comment]
From the Mozilla page (they're one of the first to implement webgpu):

"Overall, the spec is still heavily a work in progress. It’s available for early hackers but not recommended for any use in production yet. We are hoping to get a minimum-viable-product version of the spec and implementations by the end of 2020."

**Comment by JalonSolov** [comment]
As of today (almost exactly a year later), the WebGPU spec is still just a draft.

**Comment by JoshuaColell** [comment]
How about now? It's been around 3 years

**Comment by JalonSolov** [comment]
[W3C Working Draft](https://www.w3.org/standards/types#WD), 13 October 2023

So no, still not a full spec.

**Comment by medvednikov** [comment]
V supports wgpu now via Sokol.

We just upgraded to the latest Sokol, which supports wgpu.
